File: 2020 年 (25 歳) 設計、デザインパターン/design-patterns-practice/Java 言語で学ぶデザインパターン入門/01 Iterator 一 1 つ 1 つ数え上げる/02 C++/01/BookShelfIterator.cpp
#include "BookShelfIterator.h"

#include <iostream>



BookShelfIterator::BookShelfIterator(const Aggregate& aggregate) :
	Iterator(aggregate)
{
	std::cout << __FUNCTION__ << std::endl;
}

BookShelfIterator::~BookShelfIterator()
{
	std::cout << __FUNCTION__ << std::endl;
}

File: 2020 年 (25 歳) 設計、デザインパターン/design-patterns-practice/Java 言語で学ぶデザインパターン入門/01 Iterator 一 1 つ 1 つ数え上げる/01 Java/01/BookShelfIterator.java
import java.util.Iterator;



public class BookShelfIterator implements Iterator<String>
{
    private String[] elements_;
    private int position_;



    public BookShelfIterator(String[] elements_)
    {
        this.elements_ = elements_;
        this.position_ = 0;
    }

    @Override
    public boolean hasNext()
    {
        return position_ < elements_.length;
    }

    @Override
    public String next()
    {
        return elements_[position_++];
    }
}